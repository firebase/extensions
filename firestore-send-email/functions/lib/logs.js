"use strict";
/**
 * Copyright 2019 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.templateLoaded = exports.partialRegistered = exports.missingUids = exports.missingDeliveryField = exports.deliveryError = exports.delivered = exports.attemptingDelivery = exports.complete = exports.error = exports.start = exports.init = exports.obfuscatedConfig = void 0;
const config_1 = require("./config");
const firebase_functions_1 = require("firebase-functions");
exports.obfuscatedConfig = Object.assign({}, config_1.default, {
    smtpConnectionUri: "<omitted>",
});
function init() {
    firebase_functions_1.logger.log("Initializing extension with configuration", exports.obfuscatedConfig);
}
exports.init = init;
function start() {
    firebase_functions_1.logger.log("Started execution of extension with configuration", exports.obfuscatedConfig);
}
exports.start = start;
function error(err) {
    firebase_functions_1.logger.log("Unhandled error occurred during processing:", err);
}
exports.error = error;
function complete() {
    firebase_functions_1.logger.log("Completed execution of extension");
}
exports.complete = complete;
function attemptingDelivery(ref) {
    firebase_functions_1.logger.log(`Attempting delivery for message: ${ref.path}`);
}
exports.attemptingDelivery = attemptingDelivery;
function delivered(ref, info) {
    firebase_functions_1.logger.log(`Delivered message: ${ref.path} successfully. messageId: ${info.messageId} accepted: ${info.accepted.length} rejected: ${info.rejected.length} pending: ${info.pending.length}`);
}
exports.delivered = delivered;
function deliveryError(ref, e) {
    firebase_functions_1.logger.error(`Error when delivering message=${ref.path}: ${e.toString()}`);
}
exports.deliveryError = deliveryError;
function missingDeliveryField(ref) {
    firebase_functions_1.logger.error(`message=${ref.path} is missing 'delivery' field`);
}
exports.missingDeliveryField = missingDeliveryField;
function missingUids(uids) {
    firebase_functions_1.logger.log(`The following uids were provided, however a document does not exist or has no 'email' field: ${uids.join(",")}`);
}
exports.missingUids = missingUids;
function partialRegistered(name) {
    firebase_functions_1.logger.log(`registered partial '${name}'`);
}
exports.partialRegistered = partialRegistered;
function templateLoaded(name) {
    firebase_functions_1.logger.log(`loaded template '${name}'`);
}
exports.templateLoaded = templateLoaded;
