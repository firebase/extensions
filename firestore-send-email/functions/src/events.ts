import { Channel, getEventarc } from "firebase-admin/eventarc";

let eventChannel: Channel | undefined;

/** setup events */
export const setupEventChannel = () => {
  eventChannel =
    process.env.EVENTARC_CHANNEL &&
    getEventarc().channel(process.env.EVENTARC_CHANNEL, {
      allowedEventTypes: process.env.EXT_SELECTED_EVENTS,
    });
};

export const recordStartEvent = async (change) => {
  if (!eventChannel) return;

  return eventChannel.publish({
    type: "firebase.extensions.firestore-send-email.v1.onStart",
    subject: change.after.id,
    data: { doc: change.after },
  });
};

export const recordProcessingEvent = async (change) => {
  if (!eventChannel) return;

  return eventChannel.publish({
    type: "firebase.extensions.firestore-send-email.v1.onProcessing",
    subject: change.after.id,
    data: { doc: change.after },
  });
};

export const recordErrorEvent = async (change, doc, err) => {
  if (!eventChannel) return;

  return eventChannel.publish({
    type: "firebase.extensions.firestore-send-email.v1.onError",
    subject: change.after.id,
    data: { doc, err },
  });
};

export const recordSuccessEvent = async (change) => {
  if (!eventChannel) return;

  return eventChannel.publish({
    type: "firebase.extensions.firestore-send-email.v1.onSuccess",
    subject: change.after.id,
    data: { doc: change.after },
  });
};

export const recordCompleteEvent = async (change) => {
  if (!eventChannel) return;

  return eventChannel.publish({
    type: "firebase.extensions.firestore-send-email.v1.onComplete",
    subject: change.after.id,
    data: { doc: change.after },
  });
};

export const recordPendingEvent = async (change, doc) => {
  if (!eventChannel) return;

  return eventChannel.publish({
    type: "firebase.extensions.firestore-send-email.v1.onPending",
    subject: change.after.id,
    data: { doc },
  });
};

export const recordRetryEvent = async (change, doc) => {
  if (!eventChannel) return;

  return eventChannel.publish({
    type: "firebase.extensions.firestore-send-email.v1.onRetry",
    subject: change.after.id,
    data: { doc },
  });
};
