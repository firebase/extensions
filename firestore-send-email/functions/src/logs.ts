/**
 * Copyright 2019 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import config from "./config";
import { logger } from "firebase-functions";

export const obfuscatedConfig = Object.assign({}, config, {
  smtpConnectionUri: "<omitted>",
});

export function init() {
  logger.log("Initializing extension with configuration", obfuscatedConfig);
}

export function start() {
  logger.log(
    "Started execution of extension with configuration",
    obfuscatedConfig
  );
}

export function error(err: Error) {
  logger.log("Unhandled error occurred during processing:", err);
}

export function complete() {
  logger.log("Completed execution of extension");
}

export function attemptingDelivery(ref: FirebaseFirestore.DocumentReference) {
  logger.log(`Attempting delivery for message: ${ref.path}`);
}

export function delivered(
  ref: FirebaseFirestore.DocumentReference,
  info: {
    messageId: string;
    accepted: string[];
    rejected: string[];
    pending: string[];
  }
) {
  logger.log(
    `Delivered message: ${ref.path} successfully. messageId: ${
      info.messageId
    } accepted: ${info.accepted.length} rejected: ${
      info.rejected.length
    } pending: ${info.pending.length}`
  );
}

export function deliveryError(
  ref: FirebaseFirestore.DocumentReference,
  e: Error
) {
  logger.error(`Error when delivering message=${ref.path}: ${e.toString()}`);
}

export function missingDeliveryField(ref: FirebaseFirestore.DocumentReference) {
  logger.error(`message=${ref.path} is missing 'delivery' field`);
}

export function missingUids(uids: string[]) {
  logger.log(
    `The following uids were provided, however a document does not exist or has no 'email' field: ${uids.join(
      ","
    )}`
  );
}

export function partialRegistered(name) {
  logger.log(`registered partial '${name}'`);
}

export function templateLoaded(name) {
  logger.log(`loaded template '${name}'`);
}
