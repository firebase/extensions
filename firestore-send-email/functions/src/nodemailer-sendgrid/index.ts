import * as sgMail from "@sendgrid/mail";

export interface SendGridTransportOptions {
  apiKey?: string;
  [key: string]: unknown;
}

export interface Address {
  name?: string;
  address: string;
}

export interface AttachmentEntry {
  content: string | Buffer;
  filename: string;
  contentType: string;
  cid?: string;
}

export interface IcalEvent {
  content: string;
  filename?: string;
}

export interface NormalizedHeaders {
  [header: string]: string;
}

export interface MailSource {
  subject?: string;
  text?: string;
  html?: string;
  from?: Address | Address[];
  replyTo?: Address | Address[];
  to?: Address | Address[];
  cc?: Address | Address[];
  bcc?: Address | Address[];
  attachments?: AttachmentEntry[];
  alternatives?: { content: string; contentType: string }[];
  icalEvent?: IcalEvent;
  watchHtml?: string;
  normalizedHeaders?: NormalizedHeaders;
  messageId?: string;
  [key: string]: any;
  normalize(cb: (err: Error | null, source: MailSource) => void): void;
}

export class SendGridTransport {
  public readonly name: string;
  public readonly version: string;
  private readonly options: SendGridTransportOptions;

  constructor(options: SendGridTransportOptions = {}) {
    this.options = options;
    this.name = "firebase-extensions-nodemailer-sendgrid";
    this.version = "0.0.1";
    if (options.apiKey) {
      sgMail.setApiKey(options.apiKey);
    }
  }

  public send(
    mail: MailSource,
    callback: (err: Error | null, result?: unknown) => void
  ): void {
    mail.normalize((err, source) => {
      if (err) {
        return callback(err);
      }

      const msg: any = {};

      for (const key of Object.keys(source || {})) {
        switch (key) {
          case "subject":
          case "text":
          case "html":
            msg[key] = source[key];
            break;

          case "from":
          case "replyTo":
            msg[key] = []
              .concat(source[key] ?? [])
              .map((entry: Address) => ({
                name: entry.name,
                email: entry.address,
              }))
              .shift();
            break;

          case "to":
          case "cc":
          case "bcc":
            msg[key] = [].concat(source[key] ?? []).map((entry: Address) => ({
              name: entry.name,
              email: entry.address,
            }));
            break;

          case "attachments": {
            const attachments = (source.attachments ?? []).map((entry) => {
              const attachment: any = {
                content: entry.content,
                filename: entry.filename,
                type: entry.contentType,
                disposition: "attachment",
              };
              if (entry.cid) {
                attachment.content_id = entry.cid;
                attachment.disposition = "inline";
              }
              return attachment;
            });
            msg.attachments = []
              .concat(msg.attachments ?? [])
              .concat(attachments);
            break;
          }

          case "alternatives": {
            const alternatives = (source.alternatives ?? []).map((entry) => ({
              content: entry.content,
              type: entry.contentType,
            }));
            msg.content = [].concat(msg.content ?? []).concat(alternatives);
            break;
          }

          case "icalEvent": {
            const ev = source.icalEvent!;
            const attachment = {
              content: ev.content,
              filename: ev.filename ?? "invite.ics",
              type: "application/ics",
              disposition: "attachment",
            };
            msg.attachments = []
              .concat(msg.attachments ?? [])
              .concat(attachment);
            break;
          }

          case "watchHtml": {
            const alternative = {
              content: source.watchHtml!,
              type: "text/watch-html",
            };
            msg.content = [].concat(msg.content ?? []).concat(alternative);
            break;
          }

          case "normalizedHeaders":
            msg.headers = {
              ...(msg.headers ?? {}),
              ...source.normalizedHeaders,
            };
            break;

          case "messageId":
            msg.headers = {
              ...(msg.headers ?? {}),
              "message-id": source.messageId,
            };
            break;

          default:
            msg[key] = source[key];
        }
      }

      // If we've built a `content` array, merge in text/html
      if (Array.isArray(msg.content) && msg.content.length) {
        if (msg.text) {
          msg.content.unshift({ type: "text/plain", content: msg.text });
          delete msg.text;
        }
        if (msg.html) {
          msg.content.unshift({ type: "text/html", content: msg.html });
          delete msg.html;
        }
      }

      sgMail
        .send(msg)
        .then((response) => {
          callback(null, response);
        })
        .catch((err: Error) => {
          callback(err);
        });
    });
  }
}
