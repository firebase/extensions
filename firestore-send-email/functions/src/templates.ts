/**
 * Copyright 2019 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { compile } from "handlebars";

interface TemplateGroup {
  subject?: HandlebarsTemplateDelegate;
  html?: HandlebarsTemplateDelegate;
  text?: HandlebarsTemplateDelegate;
  amp?: HandlebarsTemplateDelegate;
}

export default class Templates {
  collection: FirebaseFirestore.CollectionReference;
  templateMap: { [name: string]: TemplateGroup };
  private ready: boolean;
  private waits: (() => void)[];

  constructor(collection: FirebaseFirestore.CollectionReference) {
    this.collection = collection;
    this.collection.onSnapshot(this.updateTemplates.bind(this));
    this.templateMap = {};
    this.ready = false;
    this.waits = [];
  }

  private waitUntilReady(): Promise<void> {
    if (this.ready) {
      return Promise.resolve();
    }

    return new Promise((resolve) => {
      this.waits.push(resolve);
    });
  }

  private updateTemplates(snap: FirebaseFirestore.QuerySnapshot) {
    snap.docs.forEach((doc) => {
      const data = doc.data();
      const templates: TemplateGroup = {};
      if (data.subject) {
        templates.subject = compile(data.subject, { noEscape: true });
      }
      if (data.html) {
        templates.html = compile(data.html);
      }
      if (data.text) {
        templates.text = compile(data.text, { noEscape: true });
      }
      if (data.amp) {
        templates.amp = compile(data.amp);
      }
      this.templateMap[doc.id] = templates;
      console.log(`loaded template '${doc.id}'`);
    });
    this.ready = true;
    this.waits.forEach((wait) => wait());
  }

  async render(
    name: string,
    data: any
  ): Promise<{
    subject: string | null;
    html: string | null;
    text: string | null;
    amp: string | null;
  }> {
    await this.waitUntilReady();
    if (!this.templateMap[name]) {
      return Promise.reject(
        new Error(`tried to render non-existent template '${name}'`)
      );
    }

    const t = this.templateMap[name];
    return {
      subject: t.subject ? t.subject(data) : null,
      html: t.html ? t.html(data) : null,
      text: t.text ? t.text(data) : null,
      amp: t.amp ? t.amp(data) : null,
    };
  }
}
