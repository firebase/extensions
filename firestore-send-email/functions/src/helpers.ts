import { createTransport } from "nodemailer";
import { URL } from "url";
import { invalidURI } from "./logs";
import { Config } from "./types";

function compileUrl($: string): URL | null {
  try {
    return new URL($);
  } catch (ex) {
    return null;
  }
}

export const setSmtpCredentials = (config: Config) => {
  let url;
  let transport;

  const { smtpConnectionUri, smtpPassword } = config;

  /** Generate Url object */
  url = compileUrl(smtpConnectionUri);

  /** return null if invalid url */
  if (!url) {
    invalidURI(smtpConnectionUri);

    return null;
  }

  /** encode uri password if exists */
  if (url.password) {
    url.password = encodeURIComponent(url.password);
  }

  /** encode secret password if exists */
  if (url.hostname && smtpPassword) {
    url.password = encodeURIComponent(smtpPassword);
  }

  transport = createTransport(url.href);

  return transport;
};
