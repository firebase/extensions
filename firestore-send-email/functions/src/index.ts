/**
 * Copyright 2019 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as admin from "firebase-admin";
import * as functions from "firebase-functions";
import * as nodemailer from "nodemailer";

import * as logs from "./logs";
import config from "./config";
import Templates from "./templates";
import { QueuePayload } from "./types";

logs.init();

let db;
let transport;
let templates;
let initialized = false;

/**
 * Initializes Admin SDK & SMTP connection if not already initialized.
 */
function initialize() {
  if (initialized === true) return;
  initialized = true;
  admin.initializeApp();
  db = admin.firestore();
  transport = nodemailer.createTransport(config.smtpConnectionUri);
  if (config.templatesCollection) {
    templates = new Templates(
      admin.firestore().collection(config.templatesCollection)
    );
  }
}

function validateFieldArray(field: string, array?: string[]) {
  if (!Array.isArray(array)) {
    throw new Error(`Invalid field "${field}". Expected an array of strings.`);
  }

  if (array.find((item) => typeof item !== "string")) {
    throw new Error(`Invalid field "${field}". Expected an array of strings.`);
  }
}

async function processCreate(snap: FirebaseFirestore.DocumentSnapshot) {
  // Wrapping in transaction to allow for automatic retries (#48)
  return admin.firestore().runTransaction((transaction) => {
    transaction.update(snap.ref, {
      delivery: {
        startTime: admin.firestore.FieldValue.serverTimestamp(),
        state: "PENDING",
        attempts: 0,
        error: null,
      },
    });
    return Promise.resolve();
  });
}

async function preparePayload(payload: QueuePayload): Promise<QueuePayload> {
  const { template } = payload;

  if (templates && template) {
    if (!template.name) {
      throw new Error(`Template object is missing a 'name' parameter.`);
    }

    payload.message = Object.assign(
      payload.message || {},
      await templates.render(template.name, template.data)
    );
  }

  let to: string[] = [];
  let cc: string[] = [];
  let bcc: string[] = [];

  if (typeof payload.to === "string") {
    to = [payload.to];
  } else if (payload.to) {
    validateFieldArray("to", payload.to);
    to = to.concat(payload.to);
  }

  if (typeof payload.cc === "string") {
    cc = [payload.cc];
  } else if (payload.cc) {
    validateFieldArray("cc", payload.cc);
    cc = cc.concat(payload.cc);
  }

  if (typeof payload.bcc === "string") {
    bcc = [payload.bcc];
  } else if (payload.bcc) {
    validateFieldArray("bcc", payload.bcc);
    bcc = bcc.concat(payload.bcc);
  }

  if (!payload.toUids && !payload.ccUids && !payload.bccUids) {
    payload.to = to;
    payload.cc = cc;
    payload.bcc = bcc;

    return payload;
  }

  if (!config.usersCollection) {
    throw new Error("Must specify a users collection to send using uids.");
  }

  let uids: string[] = [];

  if (payload.toUids) {
    validateFieldArray("toUids", payload.toUids);
    uids = uids.concat(payload.toUids);
  }

  if (payload.ccUids) {
    validateFieldArray("ccUids", payload.ccUids);
    uids = uids.concat(payload.ccUids);
  }

  if (payload.bccUids) {
    validateFieldArray("bccUids", payload.bccUids);
    uids = uids.concat(payload.bccUids);
  }

  const toFetch = {};
  uids.forEach((uid) => (toFetch[uid] = null));

  const documents = await db.getAll(
    ...Object.keys(toFetch).map((uid) =>
      db.collection(config.usersCollection).doc(uid)
    ),
    {
      fieldMask: ["email"],
    }
  );

  const missingUids = [];

  documents.forEach((documentSnapshot) => {
    if (documentSnapshot.exists) {
      const email = documentSnapshot.get("email");

      if (email) {
        toFetch[documentSnapshot.id] = email;
      } else {
        missingUids.push(documentSnapshot.id);
      }
    } else {
      missingUids.push(documentSnapshot.id);
    }
  });

  logs.missingUids(missingUids);

  if (payload.toUids) {
    payload.toUids.forEach((uid) => {
      const email = toFetch[uid];
      if (email) {
        to.push(email);
      }
    });
  }

  payload.to = to;

  if (payload.ccUids) {
    payload.ccUids.forEach((uid) => {
      const email = toFetch[uid];
      if (email) {
        cc.push(email);
      }
    });
  }

  payload.cc = cc;

  if (payload.bccUids) {
    payload.bccUids.forEach((uid) => {
      const email = toFetch[uid];
      if (email) {
        bcc.push(email);
      }
    });
  }

  payload.bcc = bcc;

  return payload;
}

async function deliver(
  payload: QueuePayload,
  ref: FirebaseFirestore.DocumentReference
): Promise<any> {
  logs.attemptingDelivery(ref);
  const update = {
    "delivery.attempts": admin.firestore.FieldValue.increment(1),
    "delivery.endTime": admin.firestore.FieldValue.serverTimestamp(),
    "delivery.error": null,
    "delivery.leaseExpireTime": null,
  };

  try {
    payload = await preparePayload(payload);

    if (!payload.to.length && !payload.cc.length && !payload.bcc.length) {
      throw new Error(
        "Failed to deliver email. Expected at least 1 recipient."
      );
    }

    const result = await transport.sendMail(
      Object.assign(payload.message, {
        from: payload.from || config.defaultFrom,
        replyTo: payload.replyTo || config.defaultReplyTo,
        to: payload.to,
        cc: payload.cc,
        bcc: payload.bcc,
        headers: payload.headers || {},
      })
    );
    const info = {
      messageId: result.messageId || null,
      accepted: result.accepted || [],
      rejected: result.rejected || [],
      pending: result.pending || [],
      response: result.response || null,
    };

    update["delivery.state"] = "SUCCESS";
    update["delivery.info"] = info;
    logs.delivered(ref, info);
  } catch (e) {
    update["delivery.state"] = "ERROR";
    update["delivery.error"] = e.toString();
    logs.deliveryError(ref, e);
  }

  // Wrapping in transaction to allow for automatic retries (#48)
  return admin.firestore().runTransaction((transaction) => {
    transaction.update(ref, update);
    return Promise.resolve();
  });
}

async function processWrite(change) {
  if (!change.after.exists) {
    return null;
  }

  if (!change.before.exists && change.after.exists) {
    return processCreate(change.after);
  }

  const payload = change.after.data() as QueuePayload;

  if (!payload.delivery) {
    logs.missingDeliveryField(change.after.ref);
    return null;
  }

  switch (payload.delivery.state) {
    case "SUCCESS":
    case "ERROR":
      return null;
    case "PROCESSING":
      if (payload.delivery.leaseExpireTime.toMillis() < Date.now()) {
        // Wrapping in transaction to allow for automatic retries (#48)
        return admin.firestore().runTransaction((transaction) => {
          transaction.update(change.after.ref, {
            "delivery.state": "ERROR",
            error: "Message processing lease expired.",
          });
          return Promise.resolve();
        });
      }
      return null;
    case "PENDING":
    case "RETRY":
      // Wrapping in transaction to allow for automatic retries (#48)
      await admin.firestore().runTransaction((transaction) => {
        transaction.update(change.after.ref, {
          "delivery.state": "PROCESSING",
          "delivery.leaseExpireTime": admin.firestore.Timestamp.fromMillis(
            Date.now() + 60000
          ),
        });
        return Promise.resolve();
      });
      return deliver(payload, change.after.ref);
  }
}

export const processQueue = functions.handler.firestore.document.onWrite(
  async (change) => {
    initialize();
    logs.start();
    try {
      await processWrite(change);
    } catch (err) {
      logs.error(err);
      return null;
    }
    logs.complete();
  }
);
