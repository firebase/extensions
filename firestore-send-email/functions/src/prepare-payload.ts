import { DocumentData } from "firebase-admin/firestore";
import {
  validatePayload,
  attachmentSchema,
  attachmentsSchema,
} from "./validation";
import * as logs from "./logs";
import config from "./config";
import { z } from "zod";

let db: any;
let templates: any;

export function setDependencies(database: any, templatesInstance: any) {
  db = database;
  templates = templatesInstance;
}

function validateFieldArray(field: string, array?: string[]) {
  if (!Array.isArray(array)) {
    throw new Error(`Invalid field "${field}". Expected an array of strings.`);
  }

  if (array.find((item) => typeof item !== "string")) {
    throw new Error(`Invalid field "${field}". Expected an array of strings.`);
  }
}

export async function preparePayload(
  payload: DocumentData
): Promise<DocumentData> {
  validatePayload(payload);

  const { template } = payload;

  if (templates && template) {
    if (!template.name) {
      throw new Error(`Template object is missing a 'name' parameter.`);
    }

    const templateRender = await templates.render(template.name, template.data);
    const mergeMessage = payload.message || {};

    let attachments = attachmentsSchema.parse(
      templateRender.attachments
        ? templateRender.attachments
        : mergeMessage.attachments
    );

    const handleTemplateValue = (value: any) => {
      if (value === null) {
        return undefined;
      }
      if (value === "") {
        return "";
      }
      if (value === undefined) {
        return undefined;
      }
      return value || undefined;
    };

    const templateContent = {
      subject: handleTemplateValue(templateRender.subject),
      html: handleTemplateValue(templateRender.html),
      text: handleTemplateValue(templateRender.text),
      amp: handleTemplateValue(templateRender.amp),
      attachments: attachments || [],
    };

    Object.keys(templateContent).forEach((key) => {
      if (templateContent[key] === undefined) {
        delete templateContent[key];
      }
    });

    payload.message = Object.assign(mergeMessage, templateContent);
  }

  let to: string[] = [];
  let cc: string[] = [];
  let bcc: string[] = [];

  if (typeof payload.to === "string") {
    to = [payload.to];
  } else if (payload.to) {
    validateFieldArray("to", payload.to);
    to = to.concat(payload.to);
  }

  if (typeof payload.cc === "string") {
    cc = [payload.cc];
  } else if (payload.cc) {
    validateFieldArray("cc", payload.cc);
    cc = cc.concat(payload.cc);
  }

  if (typeof payload.bcc === "string") {
    bcc = [payload.bcc];
  } else if (payload.bcc) {
    validateFieldArray("bcc", payload.bcc);
    bcc = bcc.concat(payload.bcc);
  }

  if (!payload.toUids && !payload.ccUids && !payload.bccUids) {
    payload.to = to;
    payload.cc = cc;
    payload.bcc = bcc;
    return payload;
  }

  if (!config.usersCollection) {
    throw new Error("Must specify a users collection to send using uids.");
  }

  let uids: string[] = [];

  if (payload.toUids) {
    validateFieldArray("toUids", payload.toUids);
    uids = uids.concat(payload.toUids);
  }

  if (payload.ccUids) {
    validateFieldArray("ccUids", payload.ccUids);
    uids = uids.concat(payload.ccUids);
  }

  if (payload.bccUids) {
    validateFieldArray("bccUids", payload.bccUids);
    uids = uids.concat(payload.bccUids);
  }

  const toFetch: Record<string, string | null> = {};
  uids.forEach((uid) => (toFetch[uid] = null));

  const documents = await db.getAll(
    ...Object.keys(toFetch).map((uid) =>
      db.collection(config.usersCollection).doc(uid)
    ),
    {
      fieldMask: ["email"],
    }
  );

  const missingUids: string[] = [];

  documents.forEach((documentSnapshot) => {
    if (documentSnapshot.exists) {
      const email = documentSnapshot.get("email");

      if (email) {
        toFetch[documentSnapshot.id] = email;
      } else {
        missingUids.push(documentSnapshot.id);
      }
    } else {
      missingUids.push(documentSnapshot.id);
    }
  });

  logs.missingUids(missingUids);

  if (payload.toUids) {
    payload.toUids.forEach((uid) => {
      const email = toFetch[uid];
      if (email) {
        to.push(email);
      }
    });
  }

  payload.to = to;

  if (payload.ccUids) {
    payload.ccUids.forEach((uid) => {
      const email = toFetch[uid];
      if (email) {
        cc.push(email);
      }
    });
  }

  payload.cc = cc;

  if (payload.bccUids) {
    payload.bccUids.forEach((uid) => {
      const email = toFetch[uid];
      if (email) {
        bcc.push(email);
      }
    });
  }

  payload.bcc = bcc;

  return payload;
}
