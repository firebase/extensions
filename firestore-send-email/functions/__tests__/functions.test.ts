const { logger } = require("firebase-functions");

const consoleLogSpy = jest.spyOn(logger, "log").mockImplementation();

import { obfuscatedConfig } from "../src/logs";
import * as exportedFunctions from "../src";

describe("extension", () => {
  test("functions configuration is logged on initialize", async () => {
    expect(consoleLogSpy).toBeCalledWith(
      "Initializing extension with configuration",
      obfuscatedConfig
    );
  });

  test("functions are exported", async () => {
    expect(exportedFunctions.processQueue).toBeInstanceOf(Function);
  });
});
