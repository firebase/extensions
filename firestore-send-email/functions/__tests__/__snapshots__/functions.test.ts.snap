// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extension functions configuration detected from environment variables 1`] = `
Object {
  "defaultFrom": "",
  "defaultReplyTo": "",
  "location": "europe-west2",
  "mailCollection": "",
  "smtpConnectionUri": "",
  "templatesCollection": "",
  "usersCollection": "",
}
`;
